File: test/org.eclipse.vjet.test.core.jstojava/src/org/eclipse/vjet/dsf/jst/validation/vjo/vjLib/dsf/service/DedupServiceHandler.java
/*******************************************************************************
 * Copyright (c) 2005-2012 eBay Inc.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 *******************************************************************************/

package org.eclipse.vjet.dsf.jst.validation.vjo.vjLib.dsf.service;




import java.util.List;

import org.eclipse.vjet.dsf.jsgen.shared.ids.TypeProbIds;
import org.eclipse.vjet.dsf.jsgen.shared.ids.VjoSyntaxProbIds;
import org.eclipse.vjet.dsf.jsgen.shared.validation.vjo.VjoSemanticProblem;
import org.eclipse.vjet.dsf.jst.validation.vjo.VjoValidationBaseTester;
import org.junit.Before;
import org.junit.Test;




/**
 * DedupServiceHandler.java
 * 
 * @author <a href="mailto:liama@ebay.com">liama</a>
 * @since JDK 1.5
 */
//@Category( { P3, FAST, UNIT })
//@ModuleInfo(value="DsfPrebuild",subModuleId="JsToJava")
public class DedupServiceHandler extends VjoValidationBaseTester {

    @Before
    public void setUp() {
        expectProblems.clear();
        expectProblems.add(createNewProblem2(
                VjoSyntaxProbIds.TypeUnknownNotInTypeSpace, 12, 0));
        expectProblems.add(createNewProblem2(
                VjoSyntaxProbIds.TypeUnknownNotInTypeSpace, 13, 0));
        expectProblems.add(createNewProblem2(VjoSyntaxProbIds.RedundantImport,
                14, 0));
        expectProblems.add(createNewProblem2(VjoSyntaxProbIds.RedundantImport,
                15, 0));
        expectProblems
                .add(createNewProblem2(TypeProbIds.UnusedActiveNeeds, 11, 0));
    }

    @Test
    //@Category( { P3, FAST, UNIT })
    //@Description("Test Vjo vj lib project, To validate false positive ")
    public void testDedupServiceHandler() {
        List<VjoSemanticProblem> problems = getVjoSemanticProblem(
                VjoValidationBaseTester.VJLIB_FOLDER, "vjoPro.dsf.service.",
                "DedupServiceHandler.js", this.getClass());
        assertProblemEquals(expectProblems, problems);
    }
}

File: core/test/org.eclipse.vjet.test.core.jstojava/vjLibTestFiles/vjoPro/dsf/service/DedupServiceHandler.js
/*******************************************************************************
 * Copyright (c) 2012 eBay Inc. and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     eBay Inc. - initial API and implementation
 *******************************************************************************/
vjo.ctype("vjoPro.dsf.service.DedupServiceHandler")
.needs(["vjoPro.dsf.Message",
"vjoPro.dsf.ServiceEngine",
"vjoPro.dsf.common.IJsRespHandler",
"vjoPro.dsf.common.IJsReqHandler",
"vjoPro.dsf.common.IDedupComparable"])
.satisfies("vjoPro.dsf.common.IJsRespHandler")
.satisfies("vjoPro.dsf.common.IJsReqHandler")
.protos({

messages:null,//<Object[]
oComparable:null,//<IDedupComparable
//> public void constructs(vjoPro.dsf.common.IDedupComparable);
constructs : function (comparable) {
this.messages = [];
this.oComparable = comparable;
},

//> public void handleRequest(Object);
handleRequest : function(message) {
if(!this.oComparable.shouldTrack(message)) {
return;
}
var msgs = this.messages;
for (var i = 0; i < msgs.length; i++) {
var msg = msgs[i];
if (this.oComparable.isDedup(msg,message)) {
//alert("Message is duplicate");
return;
}
}
this.messages[this.messages.length] = message;
},

//> public void handleResponse(Object);
handleResponse : function( message) {
var temp = [], len = this.messages.length;
for (var i=0; i<len; i++) {
if (message != this.messages[i]) {
temp[temp.length] = this.messages[i];
}
}
this.messages = temp;
}
})
.endType();
